### DISCUSSION TITLE
error?or whatever about Subcatchment.statistics

### DISCUSSION THREAD

**Discussion by mjzjz** [pyswmm/pyswmm#416] [Q&A] [answered]
use pyswmm Subcatchment.statistics,the total run off is about 2.35 times as much as the total runoff in Subcatchment Runoff Summary(in the .rpt file). I dont know why???the precipitation is right but the runoff and the infiltration are 2.35 times.

**Comment by bemcdonnell** [comment]
@mjzjz it would be helpful if you could provide Python/INP/RPT/OUT files. Can you zip them up and drop there here? Can you recreate it in a simple model if this is part of a bigger model?

**Reply by mjzjz** [reply]
[Uploading problem.zip…]()

**Reply by mjzjz** [reply]
can you see that problem?

**Reply by mjzjz** [reply]
[problem.zip](https://github.com/OpenWaterAnalytics/pyswmm/files/11122259/problem.zip)

**Comment by bemcdonnell** [comment, marked as answer]
@mjzjz, the units you are getting back through pyswmm are in Volume, not depth. 

https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/8266e4434bc9c23218bbf18fe57b0f3523d414ee/src/solver/stats.c#L1054-L1081

**Reply by bemcdonnell** [reply]
I think that we probably need to do a better job with units overall.  I appreciate you pointing this out!

**Reply by mjzjz** [reply]
that is helpful.Then I just divide the area by myself,or you can provide a api in convenience?

**Reply by bemcdonnell** [reply]
I am going to review some Python best practices for meta-data handling.  We could always just put this more cleanly into the documentation but that still feels like a shortcut.  We are starting out pyswmm-2 planning in the next couple weeks.  This will probably make the list.

**Reply by karosc** [reply]
I'd just use the subcatchment.area property. This discussion should prompt us to update the pyswmm docs to include units for all the data props.


for now:

```python

def depth_stats(subcatchment):
    stats = subcatchment.statistics
    area_ha = subcatchment.area       #should be area in input units (ha)
    area_sqm = area_ha  * 10_000
    mm_per_m = 1000
    for result in ['runon', 'runoff', 'evaporation', 'infiltration']:
        stats[result] = stats[result] / area_sqm * mm_per_m
    return  stats	
```

I ran the follow script and got the correct output

```python
from pyswmm import Simulation, Nodes, Links, Subcatchments
import json
def depth_stats(subcatchment):
    stats = subcatchment.statistics
    area_ha = subcatchment.area       #should be area in input units (ha)
    area_sqm = area_ha  * 10_000
    mm_per_m = 1000
    
    

    for result in ['runon', 'runoff', 'evaporation', 'infiltration']:
        stats[result] = stats[result] / area_sqm * mm_per_m
    
    return  stats

with Simulation(r'.\P=20.inp') as sim:
    ########################
    # Simulation information
    # remaining references are available here:
    # https://pyswmm.readthedocs.io/en/stable/reference/simulation.html#pyswmm.simulation.Simulation
    print("Simulation info")
    flow_units = sim.flow_units
    print("Flow Units: {}".format(flow_units))
    system_units = sim.system_units
    print("System Units: {}".format(system_units))
    print("Start Time: {}".format(sim.start_time))
    print("Start Time: {}".format(sim.end_time))

    ##################
    # Node Information
    # remaining references are available here:
    # https://pyswmm.readthedocs.io/en/stable/reference/nodes.html#pyswmm.nodes.Node

    # Node002 = Nodes(sim)["J002"]
    # print("Node 21 info")
    # print("Invert Elevation: {}".format(Node002.invert_elevation))
    # print("HOURS Surcharged: {}".format(Node002.surcharge_depth))
    # print("Physical Depth: {}".format(Node002.full_depth))
    # print("Is it a Junction?: {}".format(Node002.is_junction()))
    #
    # ##################
    # # Link Information
    # # remaining references are available here:
    # # https://pyswmm.readthedocs.io/en/stable/reference/links.html#pyswmm.links.Link
    # Link02 = Links(sim)['G00002']
    # print("Link G0002 info")
    # print("Inlet Node ID: {}".format(Link02.inlet_node))
    # print("Outlet Node ID: {}".format(Link02.outlet_node))

    subcatchments = Subcatchments(sim)

    H00000 = subcatchments['H00000']

    # Launch a simulation!
    for ind, step in enumerate(sim):
        # if ind % 100 == 0:
             pass
    #             print(sim.current_time,",", round(sim.percent_complete*100),"%", \
    #               Node009.depth, Link02.flow)
    #
    # node21_stat_out = Node002.statistics
    # print(node21_stat_out)
    # print("Max Node J002 Depth: {}".format(node21_stat_out['hours_surcharged']))

    # link15_stat_out = Link02.conduit_statistics
    # print("Link G0002 Peak Velocity: {}".format(link15_stat_out["peak_velocity"]))
    H_stat_out = H00000.statistics
    
    print("unconverted:\n",json.dumps(H_stat_out,indent=4))
    
    print("\n\nconverted:\n",json.dumps(depth_stats(H00000),indent=4))

```
 o  Retrieving project dataSimulation info
Flow Units: CMS
System Units: SI
Start Time: 2022-11-21 00:00:00
Start Time: 2022-11-21 05:00:00
unconverted:
 {
    "evaporation": 0.0,
    "infiltration": 23.527700336441118,
    "peak_runoff_rate": 0.07757799266892107,
    "precipitation": 97.81199999999998,
    "runoff": 198.82943298940467,
    "runon": 0.0
}


converted:
{
    "evaporation": 0.0,
    "infiltration": 9.923112752611186,
    "peak_runoff_rate": 0.07757799266892107,
    "precipitation": 97.81199999999998,
    "runoff": 83.85889202421116,
    "runon": 0.0
}

**Reply by mjzjz** [reply]
> I am going to review some Python best practices for meta-data handling. We could always just put this more cleanly into the documentation but that still feels like a shortcut. We are starting out pyswmm-2 planning in the next couple weeks. This will probably make the list.
that would be great!wish you success!

**Reply by mjzjz** [reply]
> I'd just use the subcatchment.area property. This discussion should prompt us to update the pyswmm docs to include units for all the data props.
> 
> for now:
> 
> ```python
> def depth_stats(subcatchment):
>     stats = subcatchment.statistics
>     area_ha = subcatchment.area       #should be area in input units (ha)
>     area_sqm = area_ha  * 10_000
>     mm_per_m = 1000
>     for result in ['runon', 'runoff', 'evaporation', 'infiltration']:
>         stats[result] = stats[result] / area_sqm * mm_per_m
>     return  stats	
> ```
> 
> I ran the follow script and got the correct output
> 
> ```python
> from pyswmm import Simulation, Nodes, Links, Subcatchments
> import json
> def depth_stats(subcatchment):
>     stats = subcatchment.statistics
>     area_ha = subcatchment.area       #should be area in input units (ha)
>     area_sqm = area_ha  * 10_000
>     mm_per_m = 1000
>     
>     
> 
>     for result in ['runon', 'runoff', 'evaporation', 'infiltration']:
>         stats[result] = stats[result] / area_sqm * mm_per_m
>     
>     return  stats
> 
> with Simulation(r'.\P=20.inp') as sim:
>     ########################
>     # Simulation information
>     # remaining references are available here:
>     # https://pyswmm.readthedocs.io/en/stable/reference/simulation.html#pyswmm.simulation.Simulation
>     print("Simulation info")
>     flow_units = sim.flow_units
>     print("Flow Units: {}".format(flow_units))
>     system_units = sim.system_units
>     print("System Units: {}".format(system_units))
>     print("Start Time: {}".format(sim.start_time))
>     print("Start Time: {}".format(sim.end_time))
> 
>     ##################
>     # Node Information
>     # remaining references are available here:
>     # https://pyswmm.readthedocs.io/en/stable/reference/nodes.html#pyswmm.nodes.Node
> 
>     # Node002 = Nodes(sim)["J002"]
>     # print("Node 21 info")
>     # print("Invert Elevation: {}".format(Node002.invert_elevation))
>     # print("HOURS Surcharged: {}".format(Node002.surcharge_depth))
>     # print("Physical Depth: {}".format(Node002.full_depth))
>     # print("Is it a Junction?: {}".format(Node002.is_junction()))
>     #
>     # ##################
>     # # Link Information
>     # # remaining references are available here:
>     # # https://pyswmm.readthedocs.io/en/stable/reference/links.html#pyswmm.links.Link
>     # Link02 = Links(sim)['G00002']
>     # print("Link G0002 info")
>     # print("Inlet Node ID: {}".format(Link02.inlet_node))
>     # print("Outlet Node ID: {}".format(Link02.outlet_node))
> 
>     subcatchments = Subcatchments(sim)
> 
>     H00000 = subcatchments['H00000']
> 
>     # Launch a simulation!
>     for ind, step in enumerate(sim):
>         # if ind % 100 == 0:
>              pass
>     #             print(sim.current_time,",", round(sim.percent_complete*100),"%", \
>     #               Node009.depth, Link02.flow)
>     #
>     # node21_stat_out = Node002.statistics
>     # print(node21_stat_out)
>     # print("Max Node J002 Depth: {}".format(node21_stat_out['hours_surcharged']))
> 
>     # link15_stat_out = Link02.conduit_statistics
>     # print("Link G0002 Peak Velocity: {}".format(link15_stat_out["peak_velocity"]))
>     H_stat_out = H00000.statistics
>     
>     print("unconverted:\n",json.dumps(H_stat_out,indent=4))
>     
>     print("\n\nconverted:\n",json.dumps(depth_stats(H00000),indent=4))
> ```
> 
> o Retrieving project dataSimulation info Flow Units: CMS System Units: SI Start Time: 2022-11-21 00:00:00 Start Time: 2022-11-21 05:00:00 unconverted: { "evaporation": 0.0, "infiltration": 23.527700336441118, "peak_runoff_rate": 0.07757799266892107, "precipitation": 97.81199999999998, "runoff": 198.82943298940467, "runon": 0.0 }
> 
> converted: { "evaporation": 0.0, "infiltration": 9.923112752611186, "peak_runoff_rate": 0.07757799266892107, "precipitation": 97.81199999999998, "runoff": 83.85889202421116, "runon": 0.0 }

thank you very much!that really helps
